src/main/java/org/kiwiproject/beta/slf4j/TimestampingLogger.java

```diff
@@ -0,0 +1,147 @@
+package org.kiwiproject.beta.slf4j;
+
+import static org.kiwiproject.base.KiwiPreconditions.requireNotNull;
+
+import com.google.common.annotations.Beta;
+import org.kiwiproject.base.KiwiStrings;
+import org.slf4j.Logger;
+import org.slf4j.event.Level;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.time.Duration;
+
+/**
+ * A simple way to log timing information about a repeated operation.
+ * <p>
+ * This is intended to be used for a single logical operation which contains multiple steps, or for the same operation
+ * repeated over a collection. For example, an order operation with separate steps to find a user, create a new order
+ * linked to the user, and insert the order to a database. Or, some kind of data migration that loops over records in
+ * a source database table and writes to a new target database table.
+ * <p>
+ * <em>Currently, this is intended only to be used within a single thread.</em>
+ */
+@NotThreadSafe
+@Beta
+public class TimestampingLogger {
+
+    @SuppressWarnings("NonConstantLogger")
+    private final Logger logger;
+    private long previousTimestamp;
+
+    /**
+     * Create a new instance using the given {@link Logger}.
+     */
+    public TimestampingLogger(Logger logger) {
+        this.logger = requireNotNull(logger);
+    }
+
+    /**
+     * Logs a message and elapsed time message at TRACE level.
+     *
+     * @param message the message or message template
+     * @param args    the arguments to the message template, if any
+     * @see #logElapsed(Level, String, Object...)
+     */
+    public void traceLogElapsed(String message, Object... args) {
```

Comment:
I feel like this and the other "logElapsed" methods should be named to be clear that it logs the given message, _and then_ logs a second time which is the elapsed time since the last time a log call was made. But I can't think of a good name, since something like `traceLogOriginalLogMessageAndThenLogElapsedTimeSinceLastLogCall` is a bit absurd - ok, it's way more than a bit absurd...

Reply:
is it ever preferable to have two log messages?

Reply:
Maybe? Like I said in my description, I grabbed this code from some code I found in a few services and "enhanced" it a bit. The original code did log the elapsed time separately, and I _think_ I did that so that I could filter only those messages in Kibana while watching our deployment and the migrations running. So, I suppose the way I actually used it in production is the "separate messages" use case.

Reply:
if the elapsedTime value was in the MDC then you could just turn it's display on and of in Kibana....

Reply:
So, would those be _separate_ methods to put the elapsed time in the MDC? e.g. `traceLogWithElapsedInMDC` or some other nonsense like that?

Reply:
no, the logger would add the MDC value to every message it logged and it would be up to the log format to include it in the output messages

Reply:
Which then forces changes to existing stuff, i.e. the log formats, right? So if all I want to do is something simple, it becomes...less simple when using MDC. Plus, IIRC the MDC uses `ThreadLocal` which in later Java versions might become problematic. I say might because I can only remember that with the addition of Virtual Threads, I thought that using `ThreadLocal` was going to be discouraged. And last, MDC uses all static methods which makes testing slightly more annoying, mostly to ensure you properly clean things up.

Reply:
I re-read the Logback [chapter](https://logback.qos.ch/manual/mdc.html) on MDCs, and MDS is mainly for _contextual_ information usually to distinguish between different clients in multi-threaded applications. Things like "client id, client's IP address, request parameters, etc". I am not sure I'd consider elapsed time information as _contextual_ information.

Reply:
possible future optional behavior

Reply:
Ok, for now we'll table it (is that correct, i.e. leave it on the table instead of doing it?)